### ISSUE TITLE
No .PDB available in static Windows builds

### ISSUE THREAD

**Issue by rustyx** [openssl/openssl#947]
It appears that during building of static Windows libs (Win32 and x64), debug symbols for both ssleay32.lib and libeay32.lib are written to the same temp file tmp32/lib.pdb. Consequently, the first lib's symbols are lost. Furthermore, the pdb is never copied to the install folder.

As a result, I'm getting a lot of linker warnings every time I try to use a static openssl lib, and if my app crashes, I can't see any symbols from openssl.

Please either fix the PDB support or (better yet) build with /Z7 flag instead of /Zi. With /Z7 the debug symbols are embedded inside the lib, just like on other platforms.

Just tested it and it works: `perl -i.bak -pe "s/\/Zi /\/Z7 /g" ms/nt.mak`

**Comment by rustyx** [comment]
In other words, the "fix" is to replace `/Zi` with `/Z7` in `ms/nt.mak`.

**Comment by levitte** [comment]
> ... Consequently, the first lib's symbols are lost.

Not quite sure how that can happen, as the .pdb file is written to during compilation.  If what you say was  possible, the file would be rewritten *at every new compilation*, which would lose all data but the last compilation unit.

> Furthermore, the pdb is never copied to the install folder.

Well, uhm, no, they aren't.  You may have noticed how they end up in the `tmp32` directory...  they were obviously never meant for public use, only for debugging OpenSSL itself.  However, actually having them in installation directories might be a good idea, something I'll ponder for upcoming 1.1.

> As a result, I'm getting a lot of linker warnings every time I try to use a static openssl lib

Would you mind demonstrating that, perhaps with a simple example?

> and if my app crashes, I can't see any symbols from openssl.

Not even for the functions you call?  That strikes me as odd...  How did you compile your app?

> With /Z7 the debug symbols are embedded inside the lib, just like on other platforms.

Actually, the trend is to have separate debugging information files when possible.  That avoids having bloated libraries.

**Comment by rustyx** [comment]
> . . . the .pdb file is written to during compilation ...

The .pdb file is written to during linking.

> Would you mind demonstrating that, perhaps with a simple example?

Here's a test project: [openssl-test.zip](http://rustyx.org/temp/openssl-test.zip). When openssl/tmp32/lib.pdb is no longer available, you will see lots of messages like:

```
1>libeay32.lib(cryptlib.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'libeay32.lib(cryptlib.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>libeay32.lib(mem.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'libeay32.lib(mem.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>libeay32.lib(mem_dbg.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'libeay32.lib(mem_dbg.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>libeay32.lib(ex_data.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'libeay32.lib(ex_data.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>ssleay32.lib(ssl_lib.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'ssleay32.lib(ssl_lib.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>ssleay32.lib(ssl_err2.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'ssleay32.lib(ssl_err2.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
1>ssleay32.lib(ssl_cert.obj) : warning LNK4099: PDB 'lib.pdb' was not found with 'ssleay32.lib(ssl_cert.obj)' or at 'C:\Projects\openssl-test\x64\Release\lib.pdb'; linking object as if no debug info
```

Of course I can manually copy tmp32/lib.pdb to the target lib folder and the warnings are gone, but still the name of the .pdb file can't be changed (so in case of openssl has to be strangely named `lib.pdb`).

Interestingly, lib.pdb seems to contain symbols from both libs, never seen this before. Even so, the tradition is to have the .pdb named the same as the module (.lib) to which it corresponds, and to have one .pdb per module (.lib, .dll or .exe).

> . . . the trend is to have separate debugging information files when possible. That avoids having bloated libraries.

Yes, for DLLs, but static libs on Windows are normally only available on a developers machine so size should not be an issue. Embedding debugging info inside static libs solves many headaches associated with .pdb location and naming.

= = =
I'm actually having difficulty reproducing the issue with the current master, as I can't find the right way to build static libs of openssl 1.1.

I tried:

```
perl Configure VC-WIN64A no-shared
nmake
nmake install
```

This generates among other things, `libcrypto.lib` and `libssl.lib`, which look like static libs, but when I try to use them I get linker errors:

```
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertOpenStore referenced in function capi_open_store
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertCloseStore referenced in function capi_find_key
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertEnumCertificatesInStore referenced in function capi_find_cert
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertFindCertificateInStore referenced in function capi_find_cert
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertDuplicateCertificateContext referenced in function capi_load_ssl_client_cert
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateContext referenced in function capi_dsa_free
1>libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertGetCertificateContextProperty referenced in function capi_cert_get_fname
```

Am I doing it wrong?

The test project can be downloaded here: [openssl-test11.zip](http://rustyx.org/temp/openssl-test11.zip).

**Comment by sav-ix** [comment]
Got similar troubles for OpenSSL builds using ICC on Windows:
`error #31000: corrupt PDB file lib\vc140.pdb; delete and rebuild; if problem persists, delete and try /Z7 instead
error code 3 ( can't write file, out of disk, etc.) opening pdb lib\vc140.pdb
compilation aborted for crypto\aes\aes_cfb.c (code 1)
NMAKE : fatal error U1077: '"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\bin\intel64\icl.EXE"' : return code '0x1'
Stop.`

The workaround is replacing `/Zi` to `/Z7` in file `Configurations/10-main.conf`. Thus +1 to
> having them in installation directories might be a good idea

**Comment by levitte** [comment]
The interesting thing here is that we have the debug information available in all cases.  In all other builds, we only have that when configuring for debugging (i.e. './config --debug')

I have no problem per se with changing `/Zi /Fdwhatever` to `/Z7`, but in that case, I'll only do that for a debug configuration.  Is that acceptable to you?  @dot-asm, a word from you about this?

**Comment by levitte** [comment]
The other option is to fine tune the use of `/Fd` a bit better, and have one PDB file for each .dll and for each .exe.  It does mean no debugging information in static libraries specifically, but...

**Comment by sav-ix** [comment]
I mean add PDB files into OpenSSL_MSVC_Debug installation directories, as some other libraries do, seems to be the best solution. And using `/Z7` instead of `/Zi` is a workaround, but not a solution.

In case of ICC, it probably require separate configuration with `/Z7` key, when and if its support would be added to OpenSSL build system.

> It does mean no debugging information in static libraries specifically

could you clarify, why MSVC_Debug_static builds could not contain PDB files too? Since OpenSSL split up MSVC_shared and MSVC_static builds, so there won't be conflicts with `.pdb` filenames.

**Comment by levitte** [comment]
> could you clarify, why MSVC_Debug_static builds could not contain PDB files too? Since OpenSSL split up MSVC_shared and MSVC_static builds, so there won't be conflicts with .pdb filenames.

PDB files are produced by LINK, and are therefore associated with DLLs and EXEs.  LIB doesn't produce any PDB, so static libraries don't have a way to associate with them.

**Comment by rustyx** [comment]
LIB does produce a PDB if linked with `/Zi` and `/DEBUG`. With `/Z7` the symbols are embedded inside the LIB. Note that specifying `/DEBUG` during _linking_ does not mean a debug build, it only means to produce symbols.

In case of static libs, whether or not to include debug symbols depends in the end on the linker setting of the target application (the one which uses the openssl libs).
It is therefore better to always include debug symbols in static libs - if no debug info is needed they will be ignored, but if debug info is requested and the symbols are not available, the linker will complain.

Pls enable debug symbols in debug and release static lib builds.

2 solutions are possible:
1. Link with `/Z7`
2. Link with `/Zi` as currently, but specify proper PDB names and copy the PDBs to the target folder.

**Comment by sav-ix** [comment]
>  LIB doesn't produce any PDB

Just made Debug_static builds of [libiconv](https://github.com/winlibs/libiconv) and [MPIR](https://github.com/wbhart/mpir) projects using MSVC 2015, and in both cases `.pdb` files were created. Thus assume, the same could be done to OpenSSL too.

**Comment by levitte** [comment]
Awright guys, can we please stop the confusion?

1.  LIB doesn't have a `/DEBUG` or a `/Zi` option.  Ref: [Overview of LIB](https://msdn.microsoft.com/en-us/library/0xb6w1f8.aspx)
2.  There are two sorts of PDB files, one produced by the compiler (`VCxx.PDF` by default) and one produced by the linker.  Ref: [PDB Files](https://msdn.microsoft.com/en-us/library/yd4f8bd1(vs.71).aspx)
3.  *We* only specify the file name for the PDB file the compiler produces, using `/Fd` (`lib.pdb` and `app.pdb` instead of the name the compiler would give it).  Ref: [/Fd (Program Database File Name)](https://msdn.microsoft.com/en-us/library/9wst99a9.aspx)
4.  *We* do *not* specify the file name for the PDB file the *linker* produces, but we *do* tell it to produce one with `/debug`.  Ref: [/PDB](https://msdn.microsoft.com/en-us/library/kwx19e36.aspx)

I just did a build to try things out, and look what I found:

```
W:\gitwrk\openssl.net\official\_build-master>dir *.pdb
 Volume in drive W is Work
 Volume Serial Number is 78EB-FB70

 Directory of W:\gitwrk\openssl.net\official\_build-master

05/12/2016  02:54 PM           233,472 app.pdb
05/12/2016  02:53 PM           503,808 lib.pdb
05/12/2016  02:52 PM         7,203,840 libcrypto-1_1.pdb
05/12/2016  02:53 PM         1,371,136 libssl-1_1.pdb
05/12/2016  02:53 PM           126,976 vc90.pdb
               5 File(s)      9,439,232 bytes
               0 Dir(s)  23,522,308,096 bytes free

W:\gitwrk\openssl.net\official\_build-master>dir *\*.pdb
The filename, directory name, or volume label syntax is incorrect.

W:\gitwrk\openssl.net\official\_build-master>dir apps\*.pdb
 Volume in drive W is Work
 Volume Serial Number is 78EB-FB70

 Directory of W:\gitwrk\openssl.net\official\_build-master\apps

05/12/2016  02:54 PM         1,739,776 openssl.pdb
               1 File(s)      1,739,776 bytes
               0 Dir(s)  23,522,308,096 bytes free

W:\gitwrk\openssl.net\official\_build-master>dir engines\*.pdb
 Volume in drive W is Work
 Volume Serial Number is 78EB-FB70

 Directory of W:\gitwrk\openssl.net\official\_build-master\engines

05/12/2016  02:53 PM           306,176 capi.pdb
05/12/2016  02:53 PM           248,832 dasync.pdb
05/12/2016  02:53 PM           232,448 ossltest.pdb
05/12/2016  02:53 PM           224,256 padlock.pdb
               4 File(s)      1,011,712 bytes
               0 Dir(s)  23,522,308,096 bytes free
```

So, please just ignore `app.pdb` and `lib.pdb`.  They are the wrong files to focus on, as they are the PDB files produced by the *compiler*

**Comment by levitte** [comment]
(I have no idea, for the moment, what that `vc90.pdb` is all about)

**Comment by sav-ix** [comment]
> (I have no idea, for the moment, what that vc90.pdb is all about)

from your link above

`Each time it creates an OBJ file, the C/C++ compiler merges debug information into VCx0.PBD. The inserted information includes type information but does not include symbol information such as function definitions.`

`The linker creates project.PDB, which contains debug information for the project's EXE file. The project.PDB contains full debug information, including function prototypes, not just the type information found in VCx0.PDB.`

Let me rephrase my previous question: what is better solution for Debug *static* builds:
- add to OpenSSL installation`.pdb` files,  with at least `type information`,
- do not add any `.pdb` files,  since they `does not include symbol information`?

**Comment by rustyx** [comment]
@levitte Thanks for clearing it up - now it is clear why the symbols from both libs are found in lib.pdb.
Sorry about the confusion.

But this does not change the fact that when I want to link with openssl statically in release mode with debug symbols, I either need openssl libs built using `/Z7` or I need lib.pdb in the same folder as the lib files. My preference is for `/Z7`, but it's your call :) I can live with a separate PDB, but then pls give it a better name and copy it together with the LIBs.

**Comment by levitte** [comment]
> > (I have no idea, for the moment, what that vc90.pdb is all about)
> 
> from your link above
> 
> `Each time it creates an OBJ file, the C/C++ compiler merges debug information into VCx0.PBD. The inserted information includes type information but does not include symbol information such as function definitions.`

That is, like I said, the *default*.  It gets overriden by the `/Fd` option...  but I did figure it out, we don't use `/Fd` when building the object files for the engines, so that's where it comes from...

> Let me rephrase my previous question: what is better solution for Debug *static* builds:
- add to OpenSSL installation`.pdb` files,  with at least `type information`,

That would be `lib.pdb`, btw.

> - do not add any `.pdb` files,  since they `does not include symbol information`?

I don't really know what's the best choice there...  I *do* know that if you're making your own program that you link with, say, `libcrypto.lib` from a non-shared configuration, then you should make sure the linker can access `lib.pdb` to get the information the linker need (see [.Pdb Files as Linker Input](https://msdn.microsoft.com/en-us/library/1d5fbft1.aspx), as far as I understand.

**Comment by levitte** [comment]
@rustyx I'll see what I can do.  Your plea makes sense :-)

**Comment by sav-ix** [comment]
> That would be lib.pdb, btw.

Such naming could be not the best solution for MSVC Debug builds. In short, it may lead to lot of warnings and debug information loss while link programs with OpenSSL. Can explain, if needed.

Thus, if it's not too late to make proposals, consider this naming scheme as option:
```
...
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\libcrypto-1_1-x64.dll
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\libcrypto-1_1-x64.pdb
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\libssl-1_1-x64.dll
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\libssl-1_1-x64.pdb
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\openssl.exe
OpenSSL-1.1.0-dev\MSVC64d\shared\bin\openssl.pdb
...
OpenSSL-1.1.0-dev\MSVC64d\shared\lib\libcrypto.lib
OpenSSL-1.1.0-dev\MSVC64d\shared\lib\libssl.lib
...

...
OpenSSL-1.1.0-dev\MSVC64d\static\bin\openssl.exe
OpenSSL-1.1.0-dev\MSVC64d\static\bin\openssl.pdb
...
OpenSSL-1.1.0-dev\MSVC64d\static\lib\libcrypto.lib
OpenSSL-1.1.0-dev\MSVC64d\static\lib\libcrypto.pdb
OpenSSL-1.1.0-dev\MSVC64d\static\lib\libssl.lib
OpenSSL-1.1.0-dev\MSVC64d\static\lib\libssl.pdb
...
```

**Comment by levitte** [comment]
It's too late for large sweeping changes, but changing something that's directly faulty is no trouble.

> Such naming could be not the best solution for MSVC Debug builds. In short, it may lead to lot of warnings and debug information loss while link programs with OpenSSL. Can explain, if needed.

I don't think the cause is in `lib.pdb`.  That file is written by the compiler, and accumulates information on all object files that were compiled with `/Fdlib`.  That includes both the libcrypto object files and the libssl ones.

**Comment by levitte** [comment]
Have a look at #1059, I believe it answers your plea

**Comment by sav-ix** [comment]
> I don't think the cause is in `lib.pdb`

Yes and no. There is no option (or I missed such) for MSVC to point *input* `.pdb` file of library, from which depends your program. E.g. you can't make MSVC use `<A name>.pdb` file for `<B name>.dll` binary.
AIUI, in part of `.pdb` files MSVC makes the decisions by itself:
- it look for `.pdb` files based on information from binaries, specified as Additional Dependencies; that binaries contain paths to build directories, which are usually obsolete, because `.pdb` files were copied to installation folders, then sources were cleaned,
- in case failure, MSVC looking for `.pdb` files in same directory as binaries, but only with names *the same* as binaries (e.g for `libcrypto-1_1-x64.dll` it look only for `libcrypto-1_1-x64.pdb`, for `libssl.lib` it would look for `libssl.pdb`),
- in case failure, MSVC generate tons of `warning LNK4099: PDB '...pdb' was not found with '...lib(...obj)' or at '...\x64\Debug\...pdb'; linking object as if no debug info`
equal to the number of missing symbols.

All above relate to some other library, but I see no difference, why this shouldn't reproduce for OpenSSL.
If I'm wrong, could you tell, how MSVC should find `lib.pdb` while linking program with `libcrypto.lib`?

Going to test PR https://github.com/openssl/openssl/pull/1059 .

**Comment by rustyx** [comment]
Looks good but how do I make a static 1.1 build? Is that described somewhere?

**Comment by levitte** [comment]
> If I'm wrong, could you tell, how MSVC should find lib.pdb while linking program with `libcrypto.lib`?

Sure (emphasis mine):

> Object (.obj) files compiled using the /Zi option contain the name of a program database (PDB). You do not specify the object's PDB file name to the linker; *LINK uses the embedded name to find the PDB if it is needed*. *This also applies to debuggable objects contained in a library*; the PDB for a debuggable library must be available to the linker along with the library.

Ref: https://msdn.microsoft.com/en-us/library/1d5fbft1.aspx

**Comment by levitte** [comment]
> Looks good but how do I make a static 1.1 build? Is that described somewhere?

`perl Configure no-shared`

You find that out by reading INSTALL

**Comment by sav-ix** [comment]
> *LINK uses the embedded name to find the PDB if it is needed*.

That's exactly what I wrote in previous message. Just checked, OpenSSL binaries contain paths:
```
[shared ]
libcrypto-1_1-x64.dll
c:\libOPENSSL-1.1.0-dev\src\libcrypto-1_1-x64.pdb 
libssl-1_1-x64.dll
c:\libOPENSSL-1.1.0-dev\src\libssl-1_1-x64.pdb  
openssl.exe
c:\libOPENSSL-1.1.0-dev\src\apps\openssl.pdb

[static]
libcrypto.lib
libssl.lib
c:\libOPENSSL-1.1.0-dev\src\lib.pdb
```

which leads to build directories. That paths became obsolete as far, as you cleaned OpenSSL sources after build. Thus LINK became unable to `uses the embedded name to find the PDB`. And the only option left if to look `.pdb` file in same directory, as depended binary with same name  as depended binary.

**Comment by levitte** [comment]
> And the only option left if to look `.pdb` file in same directory, as depended binary with same name  as depended binary.

Yes.  I assume that the issue is still about a program linked with the static libraries, right?  Then there will be a .pdb with the same name as the .exe, right?  That's the one the debugger will look for, isn't it?

**Comment by levitte** [comment]
That is, of course, as long as you linked with `/debug`

**Comment by sav-ix** [comment]
Tested PR https://github.com/openssl/openssl/pull/1059 .
For shared build everything fine, thought strange that `ossl_static.pdb` created in build directory.

For static build, would be better if there were two separate `.pdb` files in `<OpenSSL_Installation>\lib` directory (`libcrypto.pdb` for `libcrypto.lib` and `libssl.pdb` for `libssl.lib`) instead of single `ossl_static.pdb`. The reasons I mentioned above. But if that's not an option, it's not hard to make them from `ossl_static.pdb`, when needed.

As for ICC builds, its output changed from
`error #31000: corrupt PDB file lib\vc140.pdb; delete and rebuild...`
to
`error #31000: corrupt PDB file ossl_static\vc140.pdb; delete and rebuild...`
though this could be out of scope of current issue.

**Comment by levitte** [comment]
Ok, so it sounds like things do work.  `ossl_static.pdb` (a rename of `lib.pdb`) doesn't pose any *technical* problem, it only "looks strange", do I get that right?  In any case, `ossl_static.pdb` will only be  installed alonside static libraries, so with a shared build, it's not going to get installed at all.

You're right, ICC is out of scope.  It's a compiler we don't currently support.

So, except for finding my file naming strange, is there anything else, or are we done with this issue?

**Comment by sav-ix** [comment]
> I assume that the issue is still about a program linked with the static libraries, right? Then there will be a .pdb with the same name as the .exe, right? That's the one the debugger will look for, isn't it?

I told about program compilation with 3rdparty dependencies, not its debugging. And "fix" for that `warning LNK4099: PDB '...pdb' was not found ...; linking object as if no debug info` messages was to set name of required `.pdb` file the same, as corresponding binary.

> is there anything else, or are we done with this issue?

That's a question to OP. I have nothing to add to the above.

**Comment by rustyx** [comment]
It works for me now after linking with `crypt32.lib` (didn't need that in 1.0. I am guessing that's also documented somewhere;).
The PDB is nicely installed at "C:\Program Files\OpenSSL\lib\ossl_static.pdb" and I get no linker errors.
Thanks!!

**Comment by levitte** [comment]
> I told about program compilation with 3rdparty dependencies, not its debugging. And "fix" for that `warning LNK4099: PDB '...pdb' was not found ...; linking object as if no debug info` messages was to set name of required `.pdb` file the same, as corresponding binary.

I believe that to be incorrect, and did a small try.

I just made a small test...   I have my source in `w:\gitwrk\openssl.net\official\master` and my build in `w:\gitwrk\openssl.net\official\_build-master`, so I was thinking that I'd test your theory by making a `no-shared` openssl build, move libcrypto.lib and ossl_static.pdb somewhere else (so the path to the .pdb stored in the library object files wouldn't be correct) and try a very simple test program.

So, in `w:\`, I have `test.c`:

``` C
#include <stdio.h>
#include <openssl/bio.h>

int main()
{
  BIO *bio = BIO_new_fp(stdout, BIO_NOCLOSE);
  BIO_printf(bio, "%d", 1);
  BIO_free(bio);
}
```

Here's the makefile:

```
CC=cl
CFLAGS=/I gitwrk\openssl.net\official\_build-master\include /I gitwrk\openssl.net\official\master\include /MDd

all: test.exe

test.exe: test.obj
	link /nologo /debug /subsystem:console /opt:ref /out:test.exe @<<
test.obj
foo\libcrypto.lib
ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib
<<
```

Here's me moving the relevant library and .pdb:

```
W:\>md foo

W:\>move gitwrk\openssl.net\official\_build-master\ossl_static.pdb .\foo
        1 file(s) moved.

W:\>move gitwrk\openssl.net\official\_build-master\libcrypto.lib .\foo
        1 file(s) moved.

W:\>dir foo
 Volume in drive W is Work
 Volume Serial Number is 78EB-FB70

 Directory of W:\foo

05/12/2016  09:51 PM    <DIR>          .
05/12/2016  09:51 PM    <DIR>          ..
05/12/2016  07:37 PM        15,026,186 libcrypto.lib
05/12/2016  07:38 PM           520,192 ossl_static.pdb
               2 File(s)     15,546,378 bytes
               2 Dir(s)  23,508,545,536 bytes free

```

And here's the build:

```
W:\>nmake

Microsoft (R) Program Maintenance Utility Version 9.00.21022.08
Copyright (C) Microsoft Corporation.  All rights reserved.

        cl /I gitwrk\openssl.net\official\_build-master\include /I gitwrk\openss
l.net\official\master\include /MDd /c test.c
Microsoft (R) 32-bit C/C++ Optimizing Compiler Version 15.00.21022.08 for 80x86
Copyright (C) Microsoft Corporation.  All rights reserved.

test.c
        link /nologo /debug /subsystem:console /opt:ref /out:test.exe @C:\Users\
Richard\AppData\Local\Temp\nmE2EF.tmp
```

**Comment by levitte** [comment]
@rustyx, thanks!  I'm relieved.

Now, just to have someone else in the team acknowledge #1059

**Comment by levitte** [comment]
#1059 now being merged into master, it's time to close this issue.  Thank y'all!

**Comment by sav-ix** [comment]
Used your "OpenSSL-based" test of MSVC Debug builds, and it work fine. And reproduced `LNK4099` warnings for my case, which I mentioned above.
I thought, at first, that had received two mutually exclusive results. But after some investigation found that you're right. MSVC look only for PDBs, which names included in dependent binaries. It doesn't make assumptions, based on binaries names. And error was on my side, while moving installation directory with dependent binaries.
Thank you for test. It was pretty useful.
